### PR TITLE
Spring boot security

### PR DESCRIPTION
**Pull request by slizko1**

### CODE REVIEW

File: src/main/java/com/samoilenko/onlinebookstore/controller/AuthenticationController.java [hunk 1]
```diff
@@ -1,26 +1,36 @@
...
-        UserResponseDto register = userService.register(request);
-        return register;
+        return userService.register(request);
+    }
+
+    @Operation(summary = "User login", description = "Endpoint for login existing users")
+    @PostMapping("/login")
+    public boolean login(@RequestBody UserLoginRequestDto loginRequestDto) {
+        return true;
     }
```
Comment (karma-o) [inline src/main/java/com/samoilenko/onlinebookstore/controller/AuthenticationController.java]: from the task requirements:
> You don't need to implement /login endpoint in this homework!

File: src/main/java/com/samoilenko/onlinebookstore/controller/BookController.java [hunk 2]
```diff
@@ -32,32 +33,37 @@ public class BookController {
...
             description = "Returns a list of available books that can be in one page."
     )
+    @PreAuthorize("hasAnyRole('ROLE_USER', 'ROLE_ADMIN')")
     @GetMapping
     public List<BookDto> getAll(Pageable pageable) {
         return bookService.findAll(pageable);
     }
 
     @Operation(summary = "Get book by ID", description = "Returns a single book by its unique ID.")
+    @PreAuthorize("hasAnyRole('ROLE_USER', 'ROLE_ADMIN')")
```
Comment (karma-o) [inline src/main/java/com/samoilenko/onlinebookstore/controller/BookController.java]: we need to add `ROLE_` when we use `hasAuthorithy()` but here we already specified that we are checking exactly role.
Same for other similar cases
```suggestion
    @PreAuthorize("hasAnyRole('USER', 'ADMIN')")
```

File: src/main/java/com/samoilenko/onlinebookstore/dto/UserLoginRequestDto.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package com.samoilenko.onlinebookstore.dto;
+
+import jakarta.validation.constraints.Email;
+import jakarta.validation.constraints.NotBlank;
+
+public class UserLoginRequestDto {
```
Comment (karma-o) [inline src/main/java/com/samoilenko/onlinebookstore/dto/UserLoginRequestDto.java]: and lets remove this too for now

File: src/main/java/com/samoilenko/onlinebookstore/validation/constraints/FieldMatchValidator.java [hunk 4]
```diff
@@ -26,11 +28,10 @@ public boolean isValid(Object fieldsSource,
 
             Object firstValue = firstField.get(fieldsSource);
             Object secondValue = secondField.get(fieldsSource);
-
-            return firstValue == null && secondValue == null
-                    || firstValue != null && firstValue.equals(secondValue);
+            return Objects.equals(firstValue, secondValue);
         } catch (Exception e) {
-            return false;
+            throw new FieldMatchValidationException("Can't complete validation");
```
Comment (karma-o) [inline src/main/java/com/samoilenko/onlinebookstore/validation/constraints/FieldMatchValidator.java]:
```suggestion
            throw new IllegalArgumentException("Invalid field names were passed to the @FieldMatch params, validator was unable to get them or access to them, field names received: " + /*names*/);
```

File: src/main/java/com/samoilenko/onlinebookstore/model/Role.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+import jakarta.persistence.EnumType;
+import jakarta.persistence.Enumerated;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import org.springframework.security.core.GrantedAuthority;
+
+@Entity
+@Table(name = "roles")
```
Comment (karma-o) [inline src/main/java/com/samoilenko/onlinebookstore/model/Role.java]: getters/setters/equals/hashcode/toString

File: src/main/java/com/samoilenko/onlinebookstore/service/impl/UserServiceImpl.java [hunk 6]
```diff
@@ -8,20 +8,23 @@
...
     private final UserMapper userMapper;
+    private final PasswordEncoder passwordEncoder;
 
     public UserResponseDto register(UserRegistrationRequestDto requestDto)
             throws RegistrationException {
         if (userRepository.existsByEmail(requestDto.getEmail())) {
             throw new RegistrationException("User with this email already exist");
         }
         User savedUser = userMapper.toModel(requestDto);
+        savedUser.setPassword(passwordEncoder.encode(requestDto.getPassword()));
```
Comment (karma-o) [inline src/main/java/com/samoilenko/onlinebookstore/service/impl/UserServiceImpl.java:31]: we should also set default `USER` role to registered user

File: src/main/resources/db/changelog/changes/06-insert-users.yaml [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+                  value: "$2a$10$OrfBaEepFwU.OKOH/KMJ0.1IwfZVX5AnMp2iUsXxANhTIJK58ZCAO"
+              - column:
+                  name: first_name
+                  value: "Admin"
+              - column:
+                  name: last_name
+                  value: "Admin"
+              - column:
+                  name: shipping_address
+                  value: "Adminovsk"
```
Comment (karma-o) [inline src/main/resources/db/changelog/changes/06-insert-users.yaml:23]: this field is actually optional, and can be left with null value, but I like this Adminovsk more 😄

**Review by karma-o — CHANGES_REQUESTED** [review]
Several comments from me, but looks good so far!

**Review by andrii-hoienko — APPROVED** [review]
Nice!

**Pull request merged**